### PR TITLE
RPC: Add new 'clearmempool' method

### PR DESCRIPTION
**Pull request by domob1812**
The new `clearmempool` RPC method removes all transactions in the mempool.  This can be useful for testing, and might be useful also in some other situations.

### CODE REVIEW

**Comment by domob1812** [comment]
I don't know if adding this new method is desired.  I would have found it useful from time to time, but mostly related to testing (and not production use).  For instance, it can be useful in combination with `invalidateblock` to prevent the transactions from invalidated blocks from getting remined (if that is desired).

Feel free to close if you think this is not useful to have.

File: doc/release-notes.md [hunk 1]
```diff
@@ -73,6 +73,7 @@ RPC changes
 - The `fundrawtransaction` RPC will reject the previously deprecated `reserveChangeKey` option.
 - `sendmany` now shuffles outputs to improve privacy, so any previously expected behavior with regards to output ordering can no longer be relied upon.
 - The new RPC `testmempoolaccept` can be used to test acceptance of a transaction to the mempool without adding it.
+- The new RPC `clearmempool` can be used to force all mempool transactions to be removed.
```
Comment (MarcoFalke) [inline doc/release-notes.md]: If it is only for testing, it shoulnd't be mentioned in the release notes?

Comment (domob1812) [reply doc/release-notes.md]: I don't know about the exact policy and included it here to have them "complete".  But of course I'm happy to remove the mention if it is common to not include changes that are not meant (mainly) for end users.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -1924,6 +1924,28 @@ static UniValue savemempool(const JSONRPCRequest& request)
...
+        );
+    }
+
+    if (!g_is_mempool_loaded) {
+        throw JSONRPCError(RPC_MISC_ERROR, "The mempool was not loaded yet");
+    }
+
+    mempool.clear();
+
+    return NullUniValue;
```
Comment (jamesob) [inline src/rpc/blockchain.cpp]: Could return the number of evicted txns to be more useful.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -1924,6 +1924,28 @@ static UniValue savemempool(const JSONRPCRequest& request)
...
+            + HelpExampleCli("clearmempool", "")
+            + HelpExampleRpc("clearmempool", "")
+        );
+    }
+
+    if (!g_is_mempool_loaded) {
+        throw JSONRPCError(RPC_MISC_ERROR, "The mempool was not loaded yet");
+    }
+
+    mempool.clear();
```
Comment (jamesob) [inline src/rpc/blockchain.cpp]: Are we missing some lock acquisitions here (e.g. `mempool.cs`)?

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (promag) [reply src/rpc/blockchain.cpp]: Then return an object which can have more keys in the future.

File: src/rpc/blockchain.cpp (hunk 3 above)
Comment (domob1812) [reply src/rpc/blockchain.cpp]: I thought so first myself, but `CTxMempool::clear` explicitly locks for you.

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (domob1812) [reply src/rpc/blockchain.cpp]: Yes, that makes sense.  If you generally think having this method is useful, I'll add a useful return value to it (e.g. starting with the number of evicated tx).

**Comment by jonasschnelli** [comment]
Yes. Why not.
Concept ACK

File: doc/release-notes.md (hunk 1 above)
Comment (domob1812) [reply doc/release-notes.md]: Removed.

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (domob1812) [reply src/rpc/blockchain.cpp]: Changed the RPC to return a JSON object, with `nTx` holding the number of evicted transactions.

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -1924,6 +1924,37 @@ static UniValue savemempool(const JSONRPCRequest& request)
...
+static UniValue clearmempool(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 0) {
+        throw std::runtime_error(
+            "clearmempool\n"
+            "\nRemoves all transactions from the mempool.\n"
+            "It will fail if a previous dump is still being loaded.\n"
+            "\nResult:\n"
+            "{                           (json object)\n"
+            "  \"nTx\": n,                 (numeric) The number of evicted transactions\n"
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Use same key as?https://github.com/bitcoin/bitcoin/blob/990e182587999dba33f2a8172cc39d4a20d2e207/src/rpc/blockchain.cpp#L1423

Comment (domob1812) [reply src/rpc/blockchain.cpp]: Yeah, that makes sense.  (`nTx` was based on `getblock`, but `getmempoolinfo` is of course more relevant in this context.)  Will update the commit accordingly.

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -1924,6 +1924,37 @@ static UniValue savemempool(const JSONRPCRequest& request)
...
+    }
+
+    if (!g_is_mempool_loaded) {
+        throw JSONRPCError(RPC_MISC_ERROR, "The mempool was not loaded yet");
+    }
+
+    LOCK(mempool.cs);
+
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("size", mempool.size());
```
Comment (skeees) [inline src/rpc/blockchain.cpp]: maybe call this `transactions_removed` or something more informative

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -1924,6 +1924,37 @@ static UniValue savemempool(const JSONRPCRequest& request)
...
+
+    mempool._clear();
```
Comment (skeees) [inline src/rpc/blockchain.cpp]: is there a future safety benefit to calling the toplevel `mempool.clear()` here?

**Review by skeees — COMMENTED** [review]
concept ACK

I'd note that this is the first time CTxMempool::clear() will be called during normal operation (apart from startup / shutdown), so I think this requires extensive testing to make sure it leaves the mempool in a consistent state (suitable for re-use and with no cruft around).

For example, just a cursory look at that implementation suggests it might not be doing a full clear of every data structure (`mapDeltas` for example).

**Comment by MarcoFalke** [comment]
Should probably run a `CTxMemPool::check` after the clean and throw if the check fails?

**Comment by domob1812** [comment]
@MarcoFalke:  If we are worried that `CTxMemPool::clean` might leave the mempool in an inconsistent state, then we should add the `check` to `clean` (not the RPC invocation of it), right?  I don't think this is particularly necessary, but I can surely add it "just in case".  (But as I said, I think the proper place would be in `CTxMemPool::clean` and not the new RPC call.)

File: src/rpc/blockchain.cpp (hunk 5 above)
Comment (domob1812) [reply src/rpc/blockchain.cpp]: Changed.  `size` is consistent with `getmempoolinfo`, but in that context "size" is clearly more descriptive than for `clearmempool`.  I agree that here `transactions_removed` is a better name.

File: src/rpc/blockchain.cpp (hunk 6 above)
Comment (domob1812) [reply src/rpc/blockchain.cpp]: Why not - changed to `clear`.  (Especially since `_clear` does not `AssertLockHeld(cs)`.)

**Comment by domob1812** [comment]
I've now added a call to `check` to the RPC (see my previous comment - while I think the check belongs semantically into `clear`, it requires to lock `cs_main` and I think it is better to do that in the RPC than in `clear`).

`mapDeltas` is a good observation - this seems to not be caught by `check`, either.  I'll add a test for consistently between `mapDeltas` and `mapTx` there separately.

**Comment by promag** [comment]
@skeees made an excellent observation. Maybe add tests involving unconfirmed wallet transactions after clearing the mempool? Also try to include other RPC's related to the mempool, like `sendrawtransaction`, `abandontransaction` and `testmempoolaccept`?

**Comment by domob1812** [comment]
I've now split out the test into a new `mempool_clear.py` file, which I think fits better (especially if we want to have more extensive tests).  The test already has an unconfirmed wallet transaction that gets cleared from the mempool and then verifies that it is still unconfirmed after mining a block.

I'm not sure what exactly you are thinking about with other mempool-related functions.  In my opinion, the test strategy should look like this:

1. Create some "non-trivial" mempool; e.g. as now with a wallet transactions and a fee delta.
2. Call `clearmempool`.
3. Rely on `CTxMemPool::check` to verify that no inconsistencies are left (which might include extending the checker, as I did for `mapDeltas`).

I don't think that `sendrawtransaction` or `testmempoolaccept` add anything to this.  The main things to consider are that the mempool created through 1) is "complicated enough" and that `check` for step 3) is complete enough.

Do you have any concrete suggestions of what else we should do for step 1) and what might be missing for `check`?

**Comment by promag** [comment]
> I'm not sure what exactly you are thinking about with other mempool-related functions

Just a way to say that their functionality is preserved after `clearmempool`. For instance, `sendrawtransaction tx -> clearmempool -> sendrawtransaction tx` should work?

**Comment by domob1812** [comment]
@sipa:  The situation where I would have found it useful from time to time is certainly testing (not production).  You are right that restarting with wiping `mempool.dat` can reproduce the same effect, but especially the latter is at least some hassle to do.  (And honestly, when I last wanted to do it, I played around with `-nopersistmempool` but wasn't able to get it working - I didn't think about removing `mempool.dat` manually, as that seemed a bit like digging around in implementation details that shouldn't concern me as a tester.)

But if you think we should rather not expose this, then we can of course close this PR.  (And I'll just move the fix for `mapDeltas` to a separate PR without the new RPC.)

**Comment by promag** [comment]
@sipa correct me if I'm wrong (can't test ATM) but a new wallet transaction could be rejected and this would allow to add it?

**Comment by sipa** [comment]
@promag Yes, but other nodes will still reject it. Permitting is locally seems liie a bad idea in that case.

File: src/txmempool.cpp [hunk 7]
```diff
@@ -588,6 +588,7 @@ void CTxMemPool::_clear()
     mapLinks.clear();
     mapTx.clear();
     mapNextTx.clear();
+    mapDeltas.clear();
```
Comment (luke-jr) [inline src/txmempool.cpp]: `mapDeltas` probably *shouldn't* be cleared...?

File: src/txmempool.cpp [hunk 8]
```diff
@@ -727,6 +728,9 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
         assert(it2 != mapTx.end());
         assert(&tx == it->second);
     }
+    for (const auto& entry : mapDeltas) {
+        assert(mapTx.count(entry.first) > 0);
```
Comment (luke-jr) [inline src/txmempool.cpp]: Deltas aren't necessarily in the pool (needed to prioritise ones that would be rejected otherwise).

**Comment by laanwj** [comment]
>What would this RPC be useful for? It seems pretty ill advised to expose it for production use, and for testing it can always be accomplished by restarting the node (and wiping mempool.dat).

I agree with @sipa here. This has much foot-shooting potential and, besides for a few developer corner cases, doesn't seem worth the maintenance and testing hassle.

**Comment by domob1812** [comment]
Agreed - I also don't see much use besides testing, so perhaps it is safer not to add this generally.

**Pull request closed without merging**